### ISSUE TITLE
Possibly confusing error messages when using --limit with range-like selector

### ISSUE THREAD

**Issue by candlerb** [ansible/ansible#4620]
ansible 1.3.3 under both OSX and Linux. Reproducer setup:

```
==> hosts.bug <==
[storage]
storage1.example.com
storage2.example.com
storage3.example.com
storage4.example.com
storage5.example.com
storage6.example.com
storage7.example.com
storage8.example.com
storage9.example.com
dev-storage1.example.com
dev-storage2.example.com
monster3.example.com
monster4.example.com
[all:vars]
ansible_connection=local

==> bug.yml <==
- hosts: all
  gather_facts: no
  tasks:
    - action: debug msg="Talking to {{inventory_hostname}}"
```

Then run the following command:

```
$ ansible-playbook bug.yml -i hosts.bug -l 'storage[8:9].example.com'

PLAY [all] ********************************************************************

TASK: [debug msg="Talking to {{inventory_hostname}}"] *************************
ok: [storage5.example.com] => {"msg": "Talking to storage5.example.com"}

PLAY RECAP ********************************************************************
storage5.example.com       : ok=2    changed=0    unreachable=0    failed=0
```

This is rather confusing, as the inventory deals with ranges differently to -l.

Digging through code, it looks like `group[a:b]` is supposed to start at offset 'a' with limit 'b'. So I would expect `[0:2]` to talk to two hosts, but it doesn't:

```
$ ansible-playbook bug.yml -i hosts.bug -l 'storage[0:2]'

PLAY [all] ********************************************************************

TASK: [debug msg="Talking to {{inventory_hostname}}"] *************************
ok: [dev-storage1.example.com] => {"msg": "Talking to dev-storage1.example.com"}

PLAY RECAP ********************************************************************
dev-storage1.example.com   : ok=2    changed=0    unreachable=0    failed=0
```

Personally I think I'd find it more useful to be able to name a range of servers, than get an arbitrary subset of them.

Then I changed the group name `[storage]` to `[storage_servers]` in the inventory, to avoid confusion. Now the host list pattern I give results in an empty host list, which is reasonable if it's not meant to match anything:

```
$ ansible-playbook bug.yml -i hosts.bug -l 'storage[8:9].example.com'
ERROR: provided hosts list is empty
```

However, if I use a dash rather than a colon to give the range (which I thought was not allowed), I then get an exception:

```
$ ansible-playbook bug.yml -i hosts.bug -l 'storage[8-9].example.com'
  File "/Users/brian/git/ansible/lib/ansible/inventory/__init__.py", line 229, in _apply_ranges
    right=int(right)
ValueError: invalid literal for int() with base 10: '9.example.com'
```

This _is_ within code which is something to do with ranges. I don't understand what the code is doing though. A dash is treated as a special character:

```
        if "-" in rest:
            (left, right) = rest.split("-",1)
            return (first, (left, right))
        else:
            return (first, (rest, rest))
```

I did try very hard to find any documentation about the expected semantics of the `--limit` flag to ansible-playbook, but failed. It is not in:
- man ansible-playbook
- http://www.ansibleworks.com/docs/playbooks.html
- any of the pages linked in the section http://www.ansibleworks.com/docs/#special-topics-in-playbooks
- http://www.ansibleworks.com/docs/faq.html

except a mention in passing at
http://www.ansibleworks.com/docs/playbooks_checkmode.html#showing-differences-with-diff

**Comment by jimi-c** [comment]
Please see the above patch, to see if it fixes the issue for you. The documentation for this is here:

http://docs.ansible.com/playbooks_best_practices.html

which shows the proper syntax for inventory ranges is [x-y]. The ':' syntax is used as the "or" operation between groups.

**Comment by candlerb** [comment]
Thanks, I'll give the patch a try in a bit.

As regards the documentation, all I can see in that page is this example:

```
What about just the first 10, and then the next 10?:

ansible-playbook -i production webservers.yml --limit boston[0-10]
ansible-playbook -i production webservers.yml --limit boston[10-20]
```

No explanation of semantics. Is boston[10] processed twice, or are ranges non-inclusive? Does boston[0] give me the first server in the boston group, based on the order the hosts appeared in the inventory?

I also don't see any example of use of colon being used as the "or" operation between groups in that page. However I see there is an example of _semi_colon in the test case in the patch:

```
inventory.subset('greek[0-2];norse[0]')
```

There is an example of colon being used in a different way, to enumerate inventory items, at http://docs.ansible.com/intro_inventory.html

```
[webservers]
www[01:50].example.com
```

So if I wanted to have just the first 10 webservers, would I have to use
`webservers[0-10]`, which really means webservers[0] to webservers[9] inclusive, and therefore gives me www01.example.com to www10.example.com?

I'm afraid that one passing example doesn't really do this topic justice.

**Comment by candlerb** [comment]
Trying devel branch with your patch ca0ec80 applied.

```
$ ansible-playbook bug.yml -i hosts.bug -l 'storage[8:9].example.com'
ERROR: provided hosts list is empty
$ ansible-playbook bug.yml -i hosts.bug -l 'storage[0:2].example.com'
ERROR: provided hosts list is empty
$ ansible-playbook bug.yml -i hosts.bug -l 'stroage'
ERROR: provided hosts list is empty
```

That looks OK: it is what happens when any garbage pattern is passed to `--limit`

My next example was broken but strangely works:

```
$ ansible-playbook bug.yml -i hosts.bug -l 'storage[8-9].example.com'

PLAY [all] ********************************************************************

TASK: [debug msg="Talking to {{inventory_hostname}}"] *************************
ok: [storage9.example.com] => {
    "msg": "Talking to storage9.example.com"
}

$ ansible-playbook bug.yml -i hosts.bug -l 'storage[2-4].example.com'

PLAY [all] ********************************************************************

TASK: [debug msg="Talking to {{inventory_hostname}}"] *************************
ok: [storage3.example.com] => {
    "msg": "Talking to storage3.example.com"
}
ok: [storage4.example.com] => {
    "msg": "Talking to storage4.example.com"
}
```

I believe this is wrong because (as I now understand) the square-brackets are not a pattern match on the hostname, they are an indexing suffix to the group name "storage". That is, I should have written `-l 'storage[0-2]'` and `-l 'storage[2-4]'`.

Those latter versions do work as well, but I was surprised that the trailing garbage `.example.com` did not cause the pattern match to fail.

Then, if I add some more groups into the inventory:

```
[foo]
dev-storage1.example.com
dev-storage2.example.com
[bar]
monster3.example.com
monster4.example.com
```

then the following all work:

```
$ ansible-playbook bug.yml -i hosts.bug -l 'foo:bar'   # 4 hosts
$ ansible-playbook bug.yml -i hosts.bug -l 'foo;bar'   # 4 hosts
$ ansible-playbook bug.yml -i hosts.bug -l 'foo[0]:bar'   # 3 hosts
$ ansible-playbook bug.yml -i hosts.bug -l 'foo[0-1]:bar' # same 3 hosts
```

Trying out globbing:
- `-l 'fo?'` just gives hosts in group foo (fine)
- `-l 'fo??'` is empty (fine)
- `-l '???'` strangely matches all hosts. I guess it is matching the three-letter group name "all".

So, I think I can summarise the intended behaviour of the `--limit` expression as:
- Consists of one or more patterns, separated by either semicolon or colon
- Each pattern may consist of:
  - A hostname
  - A groupname
  - Hostnames and groupnames may contain glob pattern matches
    - e.g. `dev*` matches hosts and groups whose names begin with "dev"
    - "?" matches any one character and "*" matches any 0 or more characters
  - A host/group pattern with indexing suffix `[n]`, where 0 is the first host in the list
  - A host/group pattern with range suffix `[n-m]`, which gives hosts from index n to m-1 inclusive
  - Hosts and groups are expanded into a flat list before indexing is applied
    - e.g. if "foo_" matches two groups with two hosts each, then "foo_[0]" is the first host in the first group, and "foo*[3]" is the second host in the second group
  - The inventory host range syntax is NOT available as a pattern match (e.g. `www[01:50].example.com`)

Is that more or less right?

**Comment by jimi-c** [comment]
The pattern matching is always on the group name, and I don't believe is ever used on the hostname, but otherwise I think the above is correct.

**Comment by bcoca** [comment]
I use it against host names all the time, since I enforce a naming
convention by type/env/location, it is very useful.

**Comment by SupremeGit** [comment]
Unbelievable. I have hosts gs-ws01 through to gs-ws09 defined in my inventory, and yet:

ansible-playbook -i /path/to/some/inventory.yml /path/to/some/playbook.yml --limit=gs-ws0[2:3] --tags=linux-x0vnc --check
ERROR! Specified --limit does not match any hosts

trying with the x-y syntax:

ansible-playbook -i /path/to/some/inventory.yml /path/to/some/playbook.yml --limit=gs-ws0[2-3] --tags=linux-x0vnc --check
 [WARNING]: Use [x:y] inclusive subscripts instead of [x-y] which has been removed
ERROR! Specified --limit does not match any hosts

I'm seeing this completely broken behaviour on 
ansible.noarch                             2.0.1.0-2.el6               @epel
AND exactly the same after upgrade to the current ansible from epel, which is ansible 2.1.1.0-1.el6 

Somehow I seem to be the only person in the world who has noticed this.

Ansible is a joke, right? People don't actually use this garbage IRL do they?

**Comment by SupremeGit** [comment]
"The inventory host range syntax is NOT available as a pattern match (e.g. www[01:50].example.com)"

Aaaaaahhahahahahahahaha

**Comment by SupremeGit** [comment]
Great work guys, once again, you've succeeded in deliberately, and by design, making ansible as counter-intuitive and useless as possible.

Brilliant.

**Comment by mckerrj** [comment]
Supreme,  I believe you've been asked before to be less abusive to the Ansible community and team.  Accordingly, I'm afraid I'm going to ban you from commenting/posting.

**Comment by kvaps** [comment]
Hello, could you open this feature request?

I consider this issue is not solved yet. 
Commands like that still not work (ansible 2.4.2.0)
```
ansible -m ping 'node[7:11]'
ansible-playbook test.yaml -l 'node[11:23]'
``` 

regex pattern not solves this issue, because it only possible:
```
ansible -m ping 'node([7-9]|1[0:1])'
ansible-playbook test.yaml -l '~node(1[1-9]|2[0-3])'
``` 
that too complicated, and hard for understandable

**Comment by hilli** [comment]
```
→ ansible-playbook -i xamarin-hosts.prod -l devicehost152.prod,devicehost153.prod,devicehost154.prod,devicehost155.prod,devicehost156.prod,devicehost157.prod,devicehost158.prod,devicehost159.prod,devicehost160.prod,devicehost161.prod devicehosts.yaml --tags=fstrim_cron

PLAY [Devicehost configuration] *******************************************************************************************************

TASK [Gathering Facts] ****************************************************************************************************************
ok: [devicehost152.prod]
ok: [devicehost156.prod]
ok: [devicehost153.prod]
ok: [devicehost154.prod]
ok: [devicehost155.prod]

TASK [devicehost : cron] **************************************************************************************************************
changed: [devicehost155.prod]
changed: [devicehost153.prod]
changed: [devicehost152.prod]
changed: [devicehost156.prod]
changed: [devicehost154.prod]

PLAY [Devicehost configuration] *******************************************************************************************************

TASK [Gathering Facts] ****************************************************************************************************************
ok: [devicehost160.prod]
ok: [devicehost159.prod]
ok: [devicehost161.prod]
ok: [devicehost157.prod]
ok: [devicehost158.prod]

TASK [devicehost : cron] **************************************************************************************************************
changed: [devicehost161.prod]
changed: [devicehost157.prod]
changed: [devicehost159.prod]
changed: [devicehost160.prod]
changed: [devicehost158.prod]

PLAY RECAP ****************************************************************************************************************************
devicehost152.prod         : ok=2    changed=1    unreachable=0    failed=0
devicehost153.prod         : ok=2    changed=1    unreachable=0    failed=0
devicehost154.prod         : ok=2    changed=1    unreachable=0    failed=0
devicehost155.prod         : ok=2    changed=1    unreachable=0    failed=0
devicehost156.prod         : ok=2    changed=1    unreachable=0    failed=0
devicehost157.prod         : ok=2    changed=1    unreachable=0    failed=0
devicehost158.prod         : ok=2    changed=1    unreachable=0    failed=0
devicehost159.prod         : ok=2    changed=1    unreachable=0    failed=0
devicehost160.prod         : ok=2    changed=1    unreachable=0    failed=0
devicehost161.prod         : ok=2    changed=1    unreachable=0    failed=0
```

**Comment by kvaps** [comment]
Yes, for now I uses this trick:

```
ansible -m ping `echo node{7..11} | tr \  ,`
ansible-playbook test.yaml -l `echo node{11..23} | tr \  ,`
```

This is working, but ugly,  why I can't use same syntax that ansible provide in inventory file?

**Comment by ckotte** [comment]
Doesn't work with Ansible 2.5.1
```
ansible-playbook test.yml --limit hostname[31:46] --tags root_password --check
 [WARNING]: Could not match supplied host pattern, ignoring: hostname[31:46]

ERROR! Specified hosts and/or --limit does not match any hosts
```

Please implement this feature.

**Comment by kvaps** [comment]
Seems no one doesn't want to watch on this closed issue, please support my new one #42002
